src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1056,6 +1069,10 @@ protected Task ProduceEnd()
             {
                 return ProduceEndAwaited();
             }
+            else if (!HasResponseFirstFlush)
+            {
+                return ProduceEndNoStartAwaited();
```

Comment:
Instead of doing this check here and introducing ProduceEndNoStartAwaited, I would add the following to WriteSuffix:

```c#
            // ...

            if (Method == HttpMethod.Head && _responseBytesWritten > 0)
            {
                Log.ConnectionHeadResponseBodyWrite(ConnectionId, _responseBytesWritten);
            }

            if (!HasFlushedHeaders)
            {
                return FlushAsyncInternal(default);
            }

            return Task.CompletedTask;
        }

```

In the WriteSuffixAwaited case, the chunk terminator will cause the headers to be flushed anyway.

Reply:
Why would you say that's better? It seems more intuitive to be checking control flow in the same place.

Reply:
You think adding a "ProduceEndNoStartAwaited" method and conditionally calling it is simpler than just ensuring sure the headers are flushed in WriteSuffix()? I cannot imagine why. Adding "ProduceEndNoStartAwaited" adds more code and requires following what should be simple control flow through multiple methods.

Reply:
I see this is outdated. nvm

Reply:
I don't mind removing ProduceEndAwaited and moving the HasResponseStarted check up into ProduceStart if you follow my other suggestion and remove ProduceEndAwaited replacing calls with WriteSuffix.